File: src/clawcu/a2a/sidecar_plugin/openclaw/sidecar/ratelimit.js
"use strict";

// Per-peer sliding-window rate limiter for the OpenClaw A2A sidecar.
// Review-9 P2-A: the sidecar forwards /a2a/send to /v1/chat/completions,
// which triggers a provider call (LLM token spend). A misbehaving or
// compromised peer could drain budget / hit provider quota fast. This is
// a simple per-peer-name sliding-window counter — not a global defense.
//
// Design choices:
//
// - **Sliding window (not token bucket)**: we want "no more than N in the
//   last M seconds" semantics, and sliding window is dead-simple to read
//   in logs. Token bucket's "burst" concept isn't useful here — chat is
//   second-scale, a burst of 10 isn't meaningfully different from 10
//   spread over the window.
//
// - **Per-peer key**: the `from` field from /a2a/send. Mis-signed peers
//   can spoof this, but the trust model already assumes Bearer-authed
//   peers — this is quota, not auth.
//
// - **In-memory only**: sidecar is per-instance; a peer that hops between
//   instances isn't our concern. Persistence across restart would add
//   complexity for no clear gain.
//
// - **0 = disabled**: let operators opt-out via
//   A2A_RATE_LIMIT_PER_MINUTE=0. Disabled by default would make the
//   feature dead-code; enabled with a reasonable default (30/min) is
//   conservative enough that legitimate agent chat won't hit it.

function createRateLimiter({
  perMinute = 30,
  windowMs = 60 * 1000,
  nowFn = Date.now,
  maxPeers = 1024,
} = {}) {
  // peer name → array of timestamps (ms). Arrays kept small by trimming
  // on each check; maxPeers cap bounds memory if attackers rotate names.
  const hits = new Map();

  function allow(peer) {
    if (perMinute <= 0) return { ok: true, remaining: Infinity, resetMs: 0 };
    const now = nowFn();
    const windowStart = now - windowMs;
    let timestamps = hits.get(peer);
    if (!timestamps) {
      if (hits.size >= maxPeers) {
        // Evict the stalest peer to stay bounded. O(n) scan; maxPeers is
        // small enough (1024) that this is fine for sidecar traffic.
        let stalestKey = null;
        let stalestTs = Infinity;
        for (const [k, ts] of hits.entries()) {
          const last = ts[ts.length - 1] ?? 0;
          if (last < stalestTs) {
            stalestTs = last;
            stalestKey = k;
          }
        }
        if (stalestKey !== null) hits.delete(stalestKey);
      }
      timestamps = [];
      hits.set(peer, timestamps);
    }
    // Trim timestamps outside the window.
    while (timestamps.length && timestamps[0] < windowStart) {
      timestamps.shift();
    }
    if (timestamps.length >= perMinute) {
      const oldest = timestamps[0];
      const resetMs = Math.max(0, oldest + windowMs - now);
      return { ok: false, remaining: 0, resetMs };
    }
    timestamps.push(now);
    return { ok: true, remaining: perMinute - timestamps.length, resetMs: 0 };
  }

  function _peers() {
    return new Map(hits);
  }

  return { allow, _peers };
}

module.exports = { createRateLimiter };

File: src/clawcu/a2a/sidecar_plugin/openclaw/sidecar/ratelimit.py
"""Per-peer sliding-window rate limiter (Python port of ratelimit.js).

Tracks inbound /a2a/send calls per peer name. `perMinute=0` disables it. The
`max_peers` cap with stalest-first eviction keeps memory bounded against
peers that rotate their identity.
"""
from __future__ import annotations

import math
import time
from collections import deque
from dataclasses import dataclass
from typing import Callable, Deque, Dict


def _default_now_ms() -> int:
    return int(time.time() * 1000)


@dataclass
class RateDecision:
    ok: bool
    remaining: float  # math.inf when per_minute <= 0 and ok
    reset_ms: int


class RateLimiter:
    def __init__(
        self,
        per_minute: int = 30,
        window_ms: int = 60 * 1000,
        now_fn: Callable[[], int] = _default_now_ms,
        max_peers: int = 1024,
    ) -> None:
        self.per_minute = per_minute
        self.window_ms = window_ms
        self.now_fn = now_fn
        self.max_peers = max_peers
        # Insertion order preserved by dict; inner deque for O(1) popleft.
        self._hits: Dict[str, Deque[int]] = {}

    def allow(self, peer: str) -> RateDecision:
        if self.per_minute <= 0:
            return RateDecision(ok=True, remaining=math.inf, reset_ms=0)
        now = self.now_fn()
        window_start = now - self.window_ms
        timestamps = self._hits.get(peer)
        if timestamps is None:
            if len(self._hits) >= self.max_peers:
                stalest_key = None
                stalest_ts = math.inf
                for k, ts in self._hits.items():
                    last = ts[-1] if ts else 0
                    if last < stalest_ts:
                        stalest_ts = last
                        stalest_key = k
                if stalest_key is not None:
                    self._hits.pop(stalest_key, None)
            timestamps = deque()
            self._hits[peer] = timestamps
        while timestamps and timestamps[0] < window_start:
            timestamps.popleft()
        if len(timestamps) >= self.per_minute:
            oldest = timestamps[0]
            reset_ms = max(0, oldest + self.window_ms - now)
            return RateDecision(ok=False, remaining=0, reset_ms=reset_ms)
        timestamps.append(now)
        return RateDecision(ok=True, remaining=self.per_minute - len(timestamps), reset_ms=0)

    def _peers(self) -> Dict[str, Deque[int]]:
        # Shallow copy for tests/inspection, mirroring the Node helper.
        return {k: deque(v) for k, v in self._hits.items()}


def create_rate_limiter(
    per_minute: int = 30,
    window_ms: int = 60 * 1000,
    now_fn: Callable[[], int] = _default_now_ms,
    max_peers: int = 1024,
) -> RateLimiter:
    return RateLimiter(per_minute=per_minute, window_ms=window_ms, now_fn=now_fn, max_peers=max_peers)
